_posts/2019-01-24-how-to-keep-safe-as-a-client.md

```diff
@@ -0,0 +1,130 @@
+- Percentage, for example, about 30% of request can't get response or have a latency.
+
+We can use pseudo code to express above definitions:
+
+```python
+errorCount = 0;
+isOpenCircuitBreaker = false;
+limitCount = 10;
+
+if(success){
+  return success;
+}else{
+  errorCount++;
+  if(errorCount >= limitCount){
+    isOpenCircuitBreaker = true;
+  }
+}
+
+```
+
+### Cut off connection
+
+Once our downstram system is broken down, we need to cut off connection instantly, This action can both reduce useless remote call and reduce downstream system's load.
+for example, we use RPC framework to communicate between systems.
+![](https://mmbiz.qpic.cn/mmbiz_png/oB5bd6W6hI31YcnTS8xgHlND7GJnk9vS47MgWD0aWMEqAYpvxibphTppSvBia6tRo3wpDgBgTKys8MqtgzwcqWXw/640?wx_fmt=png&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1)
+ServiceA is a client, If we find ServiceB is unavailable, we can return failed response within ServiceA and do not invoke ServiceB
+![](https://mmbiz.qpic.cn/mmbiz_png/oB5bd6W6hI31YcnTS8xgHlND7GJnk9vS7HACeboGiaUlzjRXWJhR2nGMJQIYkfAxZE1TRfTIXME68kC6YPmRaIg/640?wx_fmt=png&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1)
+This is so called "Fail Fast". we can use below code to implement Fail Fast.
+
+```python
+if(isOpenCircuitBreaker == true){
+  return fail;
+}
+
+// we invoke remote service here
+
+```
+### Try to reconnect
+
+After cut off connections, We need to recover as soon as possbile, Because currently the system is not complete. In order to automatically recover connection we need to define a standard, like previously how we judge a system is unavailable, we need to define whether a system is abaliable again.
+- Threshold, for example, there are 100 success request within 5 seconds.
+- Percentage, for example, there are 95% success request within 10 seconds.
+
+It also contains time window, threshold and percentage. But there is a slight difference: most of the time, the system will keep unavailable for a momment, So we can check in a fixed times(30s for example).
+
+```python
+successCount = 0;
+isHalfOpen = true;
+
+if(success){
+  if(isHalfOpen){
+    successCount++;
+    if(successCount == threshold){
+      isOpenCircuitBreaker = false;
+    }
+  }
+  return success;
+}else{
+  errorCount++;
+  if(errorCount == threshold){
+    isOpenCircuitBreaker = true;
+  }
+}
+
+```
+However reconnect may failed, In a large system we can't use all servers to try reconnecting. We can take these measures:
+- 1.Use a few servers to check
+- 2.Add a health check interface to judge system conditions: CPU, IO for example.
+
+### recover
+
+Once the system is avaliable, we need to recover.
+
+![](https://mmbiz.qpic.cn/mmbiz_png/oB5bd6W6hI31YcnTS8xgHlND7GJnk9vSUGPwicRFU6xJeNPxzKOWEhlM3yKSVdEfk1nusm5FfnRm1WyYkOVybBg/640?wx_fmt=png&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1)
+
+This is the general process of Circuit Breaker, Frameworks like Hystrix, resilience4j are all have such process flow.
```

Comment:
... are all have such process flow -> ... are designed based on such process flow (concept)

Reply:
Thanks alvita! you are so patient and careful! I updated these spelling mistakes and add a reference article!